crates/ruff_python_parser/src/lexer.rs

```diff
@@ -428,10 +429,46 @@ impl<'source> Lexer<'source> {
                     self.cursor.bump();
                     value.push('\\');
                 }
+                '?' => {
+                    self.cursor.bump();
+                    let mut question_count: u8 = 1;
+                    while self.cursor.eat_char('?') {
+                        question_count += 1;
+                    }
+
+                    // The help end magic command is valid if it has 1 or 2 question marks.
+                    // If the `value` is empty i.e., we are at the start of the line,
+                    // then we can't be sure if it is a help end magic command or not.
+                    if question_count > 2
+                        || previous_char_is_whitespace
+                        || value.is_empty()
+                        || !matches!(self.cursor.first(), '\n' | '\r' | EOF_CHAR)
```

Comment:
I would look at the last character in `value` to avoid having to write a state variable with every character (in addition to updating the vector).


```suggestion
                        || value.last().is_some_and(|c| !is_python_whitespace(c))
                        || !matches!(self.cursor.first(), '\n' | '\r' | EOF_CHAR)
```